types/jasmine/ts3.1/index.d.ts

```diff
@@ -252,44 +265,38 @@ declare namespace jasmine {
 
     function addMatchers(matchers: CustomMatcherFactories): void;
 
-    function stringMatching(str: string | RegExp): Any;
+    function stringMatching(str: string | RegExp): AsymmetricMatcher<string>;
 
     function formatErrorMsg(domain: string, usage: string): (msg: string) => string;
 
-    interface Any {
-        (...params: any[]): any; // jasmine.Any can also be a function
-        new (expectedClass: any): any;
-
-        jasmineMatches(other: any): boolean;
-        jasmineToString(): string;
-    }
-
-    interface AsymmetricMatcher<T extends string = string> {
-      asymmetricMatch(other: any): boolean;
-      jasmineToString?(): T;
+    interface AsymmetricMatcher<TValue> {
+        asymmetricMatch(other: TValue): boolean;
+        // Mark it optional to support custom user matchers coming without describe function.
+        jasmineToString?(): string;
     }
 
-    interface Truthy extends AsymmetricMatcher<'<jasmine.truthy>'> { }
-    interface Falsy extends AsymmetricMatcher<'<jasmine.falsy>'> { }
-    interface Empty extends AsymmetricMatcher<'<jasmine.empty>'> { }
-    interface NotEmpty extends AsymmetricMatcher<'<jasmine.notEmpty>'> { }
-
     // taken from TypeScript lib.core.es6.d.ts, applicable to CustomMatchers.contains()
     interface ArrayLike<T> {
         length: number;
         [n: number]: T;
     }
 
-    interface ArrayContaining<T> extends AsymmetricMatcher {
-        new?(sample: ArrayLike<T>): ArrayLike<T>;
-    }
+    interface ArrayContaining<T> extends AsymmetricMatcher<ArrayLike<T>> { }
 
-    interface ObjectContaining<T> {
-        new?(sample: {[K in keyof T]?: any}): {[K in keyof T]?: any};
+    /**
+     * @deprecated Use AsymmetricMatcher<any> instead.
+     * @see {AsymmetricMatcher<any>}
+     */
+    type Any = AsymmetricMatcher<any>;
 
-        jasmineMatches(other: any, mismatchKeys: any[], mismatchValues: any[]): boolean;
-        jasmineToString?(): string;
-    }
+    type UnwrapAsymmetricMatchers<T> = {
+        [K in keyof T]: T[K] extends AsymmetricMatcher<infer TMatching> ? TMatching : UnwrapAsymmetricMatchers<T[K]>
+    };
+
+    // Current matcher reports the compatible type (type it can match) via AsymmetricMatcher arg.
+    // If any our property is a matcher itself, we should unwrap it, so TS could verify
+    // whether corresponding object property can be actually matched by that matcher.
+    interface ObjectContaining<T> extends AsymmetricMatcher<UnwrapAsymmetricMatchers<T>> { }
```

Comment:
Thanks for this PR! I have a problem wrapping my head around this — does this actually work with deeply nested `objectContaining` calls?

The current typings don't support this, forcing us to cast to `any`. The test cases suggest that it might work, but it might be worth to add an additional test case that tests this explicitly.

To be clear, I'm talking about a case like this:
```ts
expect(foo).toEqual(
  jasmine.objectContaining({
    someObject: jasmine.objectContaining({
      that: jasmine.objectContaining({
        isReally: jasmine.arrayContaining([
          jasmine.objectContaining({
            deeply: 'nested',
          },
        ]),
      }),
    }),
  })
);
```

Reply:
Hi, and thanks for the feedback!

Yes, It's fully supported now - see [this correct check](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38087/files#diff-c6dace63638c5593ee88c4ec90b42cc9R1624) and [this failure](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38087/files#diff-c6dace63638c5593ee88c4ec90b42cc9R1637) 😊

Reply:
Awesome, thank you!